File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    ordered_base = order(base)
    Enum.filter(candidates, fn(candidate) ->
      ordered_base == order(candidate) && String.downcase(base) != String.downcase(candidate)
    end)
  end

  defp order(string) do
    string
    |> String.downcase
    |> String.to_char_list
    |> Enum.sort
    |> Enum.join
  end
end

File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))

(defn- anagram?
  [word candidate]
  (=
    (sort (str/lower-case word))
    (sort (str/lower-case candidate))))

(defn- remove-from-list
  [word words]
  (remove #(= (str/lower-case word) (str/lower-case %)) words))

(defn anagrams-for
  [word candidates]
  (->>
    candidates
    (remove-from-list word)
    (filter (fn [candidate]
      (anagram? word candidate)))))
